src/rt/cover.d

```diff
@@ -149,10 +149,83 @@ uint parseNum(const(char)[] s)
 T min(T)(T a, T b) { return a < b ? a : b; }
 T max(T)(T a, T b) { return b < a ? a : b; }
 
+version (unittest) __gshared bool inUnittest;
```

Comment:
Meh this global is truly ugly. Why not make it a parameter to the function instead ?

Reply:
Also taken from `src/gc/config.d`. I guess in this case an argument can be passed, although to be honest, if the problem here is ugliness, I don't think passing an extra pointless argument in every call to `parseMergeOpts` will add any beauty.

Reply:
You could avoid it completely and then just in the unittests do this:

```
   assert(parseMergeOpt("yEs")[0.."DRT Error".length] == "DRT Error" && !merge);
```

Reply:
Whoops, that's wrong, since you're printing to `stderr`, not returning this.

Reply:
Yep that was my first thought as well. But here you are declaring a public global used as a flag for a single function from a single calling point. How else can I say that a global is one of the worst way to do it, unless you're a C programmer from the 60's ?

Reply:
Why is the worst way to do it. I don't see any real technical reason. Just saying something is ugly or old is not good enough for a reason to me. Tell me why is it bad.

Reply:
To explain a bit more, I know many ways in which globals could be bad. I will tell you why I think a global is good in this case:

* Avoids polluting the function signature with something that's only useful for unit testing.
* It gets completely eliminated when not unit testing, not only the global but all the code that is used for unit testing.
* It is completely encapsulated, is private to the module and used only when unit testing.

Reply:
Would it be better if you marked this global `private`?

Reply:
> It is completely encapsulated, is private to the module and used only when unit testing.
 
I think this is wrong.

Reply:
>  I think this is wrong.

Actually there's a `private:` a couple of lignes before so it's okay-ish.

My reasons mainly boils down to SoC. You mention:

> Avoids polluting the function signature with something that's only useful for unit testing.

But this information *actually* belongs to the function signature ! Because it only affects that function, and not the whole module. When I see a global in a module, I expect it to be used in several places, and I truly hope it's read-only (or at least written-once) because that complicates understanding of the code a great deal.

If you really want something clean, make it take a sink which either does nothing or forward to `fprintf`.

> It gets completely eliminated when not unit testing, not only the global but all the code that is used for unit testing.

Good point though, but does it really matter in that case ?

If you feel strongly about it, I'm totally okay to drop the ball, I was just really surprised by this code.

Reply:
I'm a bit indifferent, I don't care much either way, I'm just lazy and repeating a practice I already found in the code. But I do think is a simple solution, I don't think it complicates reading the code, is super straight-forward to read `inUnittest` and know what is it about and when is true and when it's false. When I see it in the original code I didn't see anything wrong and it was perfectly clear to me.

I'll wait to see what the druntime maintainers think and change it only they prefer it gone.

Reply:
> I'll wait to see what the druntime maintainers think and change it only they prefer it gone.

 Another solution would be to define a specific version which is only set during druntime's unittest compilation

Reply:
This is not good enough, because then other tests hitting this code will print to stderr. the version should be only set while this particular unittest is running. AFAIK that can't be done with a `version`, right?

Reply:
Yes, it can't, but is there any other bit of unittest code at Druntime that should `printf` instead of "err"?

Reply:
It's a common pattern, e.g. see cpuid.